File: iosApp/iosApp/Services/Etherscan/EtherscanService.swift
// Created by web3d4v on 05/08/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3lib

struct EtherscanTransactionResponse: Codable {
    let status: String
    let message: String
    let result: [EtherscanTransaction]
}

struct EtherscanTransaction: Codable {
    let blockNumber: String
    let timeStamp: String
    let hash: String
    let nonce: String
    let blockHash: String
    let transactionIndex: String
    let from: String
    let to: String
    let value: String
    let gas: String
    let gasPrice: String
    let isError: String
    let txreceiptStatus: String
    let input: String
    let contractAddress: String
    let cumulativeGasUsed: String
    let gasUsed: String
    let confirmations: String
    let methodId: String
    let functionName: String
}

enum EtherscanServiceError: Error {
    
    case unableToConstructURL
    case failedToDownload
}

protocol EtherscanService {

    func cachedTransactionHistory(
        for walletAddress: String,
        nonce: String
    ) -> [EtherscanTransaction]
    
    func fetchTransactionHistory(
        for address: String,
        onCompletion: @escaping (Result<[EtherscanTransaction], Error>) -> Void
    )
}

final class DefaultEtherscanService {
    
    private let networksService: NetworksService
    private let defaults: UserDefaults
    
    private let LATEST_FILE_NAME = "EtherScanFileNameCached"
    // TODO: @Anoon to provide API KEY
    private let API_KEY = ""
    
    init(
        networksService: NetworksService,
        defaults: UserDefaults = .standard
    ) {
        self.networksService = networksService
        self.defaults = defaults
    }
}

extension DefaultEtherscanService: EtherscanService {
    
    func cachedTransactionHistory(
        for walletAddress: String,
        nonce: String
    ) -> [EtherscanTransaction] {
        
        let transactionFileTarget = makeFileName(for: walletAddress, nonce: nonce)
        let transactionFileCached = defaults.string(forKey: LATEST_FILE_NAME)
        
        guard transactionFileTarget == transactionFileCached else {
            clearCache()
            return []
        }
        
        guard let data = defaults.data(forKey: transactionFileTarget) else {
            clearCache()
            return []
        }
        
        guard let transactions = try? loadTransactions(from: data) else {
            clearCache()
            return []
        }
        
        return transactions
    }
    
    func fetchTransactionHistory(
        for walletAddress: String,
        onCompletion: @escaping (Result<[EtherscanTransaction], Error>) -> Void
    ) {
        
        guard let urlRequest = makeURLRequest(for: .transactions(address: walletAddress)) else {
            onCompletion(.failure(EtherscanServiceError.unableToConstructURL))
            return
        }
        
        URLSession.shared.dataTask(with: urlRequest) { [weak self] data, _, error in
            guard let self = self else { return }
            guard let data = data else {
                onCompletion(.failure(error ?? EtherscanServiceError.failedToDownload))
                return
            }

            do {
                let decoder = JSONDecoder()
                decoder.keyDecodingStrategy = .convertFromSnakeCase
                let transactions = try decoder.decode(
                    EtherscanTransactionResponse.self,
                    from: data
                ).result
                
                try self.storeNewTransactions(with: walletAddress, and: transactions)
                
                DispatchQueue.main.async {
                    onCompletion(.success(transactions))
                }
            } catch {
                DispatchQueue.main.async {
                    onCompletion(.failure(error))
                }
            }
        }.resume()
    }
}

private extension DefaultEtherscanService {
    
    func clearCache() {
        guard let transactionFileCached = defaults.string(forKey: LATEST_FILE_NAME) else {
            
            return
        }
        defaults.removeObject(forKey: transactionFileCached)
        defaults.synchronize()
    }
    
    func makeFileName(for wallet: String, nonce: String) -> String {
        "\(wallet)_\(nonce)"
    }
    
    func storeNewTransactions(
        with walletAddress: String,
        and transactions: [EtherscanTransaction]
    ) throws {
        
        guard let nonce = transactions.first?.nonce else { return }
        
        let fileName = makeFileName(for: walletAddress, nonce: nonce)
        
        let encoder = JSONEncoder()
        let data = try encoder.encode(transactions)
        
        clearCache()
        
        defaults.set(fileName, forKey: LATEST_FILE_NAME)
        defaults.set(data, forKey: fileName)
        defaults.synchronize()
    }
    
    func loadTransactions(from data: Data) throws -> [EtherscanTransaction] {
        
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        return try decoder.decode(
            [EtherscanTransaction].self,
            from: data
        )
    }
}

private extension DefaultEtherscanService {
    
    var host: String { "api.etherscan.io" }
    
    enum Details {
        case transactions(
            address: String
        )
        
        var path: String {
            switch self {
            case .transactions:
                return "/api"
            }
        }
    }
    
    func makeURLRequest(
        for details: Details
    ) -> URLRequest? {
        
        var urlComponents = URLComponents()
        urlComponents.scheme = "https"
        urlComponents.host = host
        
        switch details {
        case let .transactions(address):
            urlComponents.path = details.path
            urlComponents.queryItems = [
                .init(name: "module", value: "account"),
                .init(name: "action", value: "txlist"),
                .init(name: "address", value: address),
                .init(name: "sort", value: "desc")
            ]
            if !API_KEY.isEmpty {
                urlComponents.queryItems?.append(
                    .init(name: "apikey", value: API_KEY)
                )
            }
        }
        
        guard let url = urlComponents.url else { return nil }
        
        var urlRequest = URLRequest(url: url)
        urlRequest.allHTTPHeaderFields = [
            "Content-Type": "application/json",
            "Accept": "application/json"
        ]
        return urlRequest
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/services/etherScan/EtherScanService.kt
package com.sonsofcrypto.web3walletcore.services.etherScan

import com.sonsofcrypto.web3lib.BuildKonfig
import com.sonsofcrypto.web3lib.utils.KeyValueStore
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.extensions.jsonDecode
import com.sonsofcrypto.web3lib.extensions.jsonEncode
import com.sonsofcrypto.web3lib.extensions.stdJson
import io.ktor.client.HttpClient
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.plugins.logging.LogLevel
import io.ktor.client.plugins.logging.Logger
import io.ktor.client.plugins.logging.Logging
import io.ktor.client.plugins.logging.SIMPLE
import io.ktor.client.request.get
import io.ktor.client.request.headers
import io.ktor.client.statement.bodyAsText
import io.ktor.http.ContentType
import io.ktor.http.HttpHeaders
import io.ktor.http.URLProtocol
import io.ktor.http.path
import io.ktor.http.withCharset
import io.ktor.serialization.kotlinx.json.json
import io.ktor.utils.io.charsets.Charsets
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
private data class EtherScanResponse(
    val status: String,
    val message: String,
    val result: List<EtherScanTransaction>,
)

@Serializable
data class EtherScanTransaction(
    val blockNumber: String,
    val timeStamp: String,
    val hash: String,
    val nonce: String,
    val blockHash: String,
    val transactionIndex: String,
    val from: String,
    val to: String,
    val value: String,
    val gas: String,
    val gasPrice: String,
    val isError: String,
    @SerialName("txreceipt_status")
    val txreceiptStatus: String,
    val input: String,
    val contractAddress: String,
    val cumulativeGasUsed: String,
    val gasUsed: String,
    val confirmations: String,
    val methodId: String,
    val functionName: String,
)

interface EtherScanService {
    /** Return local cached transaction history for the given address, network and nonce */
    fun transactionHistory(
        address: String, network: Network, nonce: String
    ): List<EtherScanTransaction>

    @Throws(Throwable::class)
    /** Update the local cached transaction history for the given address and network */
    suspend fun fetchTransactionHistory(address: String, network: Network)
}

class DefaultEtherScanService(
    private val store: KeyValueStore,
): EtherScanService {

    private val client: HttpClient by lazy {
        HttpClient {
            Logging { level = LogLevel.NONE; logger = Logger.SIMPLE }
            install(ContentNegotiation) {
                json(
                    stdJson, ContentType.Application.Json.withCharset(
                        Charsets.UTF_8
                    )
                )
            }
        }
    }

    override fun transactionHistory(
        address: String,
        network: Network,
        nonce: String
    ): List<EtherScanTransaction> {
        val key = storageKey(address, network, nonce)
        if (key != transactionsKey) {
            clearCachedTransactions()
        }
        val transactions = cachedTransactions()
        if (transactions.isEmpty()) {
            clearCachedTransactions()
        }
        return transactions
    }

    override suspend fun fetchTransactionHistory(address: String, network: Network) {
        val body = client.get {
            headers {
                append(HttpHeaders.Accept, "application/json")
            }
            url {
                protocol = URLProtocol.HTTPS
                host = network.host
                path("api")
                parameters.append("module", "account")
                parameters.append("action", "txlist")
                parameters.append("address", address)
                parameters.append("sort", "desc")
                parameters.append("apikey", BuildKonfig.etherscanKey)
            }
        }.bodyAsText()

        val transactions = jsonDecode<EtherScanResponse>(body)?.result ?: emptyList()
        storeTransactions(address, network, transactions)
    }

    private fun clearCachedTransactions() {
        store[cachedKey] = null
        store[transactionsKey] = null
    }

    private fun cachedTransactions(): List<EtherScanTransaction> =
        jsonDecode(store[transactionsKey] ?: "[]") ?: emptyList()

    private fun storageKey(address: String, network: Network, nonce: String): String {
        return "${address}_${network.id()}"
    }
    private val cachedKey: String = "EtherScanFileNameCached"
    private val transactionsKey: String get() = store[cachedKey] ?: ""
    private fun storeTransactions(
        address: String,
        network: Network,
        transactions: List<EtherScanTransaction>
    ) {
        val nonce = transactions.firstOrNull()?.nonce?.toInt() ?: return
        val storageKey = storageKey(address, network, (nonce + 1).toString())
        clearCachedTransactions()
        store[cachedKey] = storageKey
        store[transactionsKey] = jsonEncode(transactions)
        println("1 ----------")
        println("Storing: $storageKey")
        println("Transactions: ${transactions.count()}")
        println("After Storing:")
        val list = cachedTransactions()
        println("2 --- cached(${list.count()})-------")
    }

    private val Network.host: String get() {
        return when (this.chainId) {
            3u -> "api-ropsten.etherscan.io"
            4u -> "api-rinkeby.etherscan.io"
            5u -> "api-goerli.etherscan.io"
            else ->"api.etherscan.io"
        }
    }
}